### ISSUE TITLE
Running to-do

### ISSUE THREAD

**Issue by eshedmargalit** [neuroailab/spacenet#11]
### Things that could be wrong
1. Spatial distribution of units
    1a. Start with uniform density disc
    1b. Scaling of spatial extent (i.e., density)

2. Optimization
   2a. Try bumping out of local minima (add outer loop that injects noise at convergence)
   2b. Other ways to decrease loss function further

3. Feature distribution of V1 units
    3a. Cardinal orientation bias
    3b. Compare to published data

4. Loss function

5. Image set used to derive features for optimization

Working on 3-5 likely require getting 1-2 fixed first

**Comment by eshedmargalit** [comment]
Validation set: highest spatial frequency likely needs to be set even higher
1. Some high SF filters are being assigned low SF images as best fits, potentially because there isn't an SF high enough to match their RF
2. Many OS filters are showing ramp tuning, but we haven't found the point where tuning starts to decrease

**Comment by eshedmargalit** [comment]
# 1/23

### Isolating small-distance within-peak case where image correlations are high
1. When do swaps occur? How does swap occurrence depend on density of units?
2. How correlated are units within these regime (960 x 960 matrix)?

### Feature distribution of V1 units
1. Cardinal orientation bias (done)
2. Compare to published data (unsure which data to look at)

**Comment by eshedmargalit** [comment]
1. Spacenet training at "correct" range
2. "correct" range but actually sampling points from throughout the retinotopic map; only compute delta_L for the two entries that swap to make it tractable
3. Consider normalizations for range of Cs and range of d's

**Comment by eshedmargalit** [comment]
#2/7/19
1. Full map, local subsampling
2. OS-only at first, then expand to all filters

**Comment by eshedmargalit** [comment]
# 2/15/19

1. Try alternative loss function, e.g., `C^2` instead of `|C|`
1. Map smoothing
1. Contour lines
1. Evaluate orthogonality
1. Reading about pinwheel quantification metrics

**Comment by eshedmargalit** [comment]
Restore implemented simply; load object from pkl and keep swapping
This is fairly fragile because it depends on the signature of `FeatureMapRF` not changing much, so in the future it'd be smarter to instantiate a new object and overwrite the positions with the starting positions from the restore point. In fact, the right way to do it would be to

1. Create a new FM, providing subsample indices from fm_restore.subsampled["indices"]
2. Overwrite positions from fm_restore.pos{x, y}

Okay, so doing this likely means offering restore params somewhere in __init__
Dictionary like this might work:

```
restore_params : {
    'indices': fm_restore.subsampled["indices"],
    'posx': fm_restore.posx,
    'posxy': fm_restore.posy,
}
```

**Comment by eshedmargalit** [comment]
# 02/28/19

1. Expand test set for spatial frequencies
2. Fix clusterness metrics
    1. Use same bins as Nauhaus
    2. Fit Gaussians to tuning curves
3. Orthogonality metrics
4. More rigorously nail down automatic distance term scaling
5. Map autocorrelations (to compare spatial scale/periodicity of maps)
6. TPU subsampling

**Comment by eshedmargalit** [comment]
# 03/07/19
1. Only non-OS (pinwheels for color?) swap-opt
2. Swapopt with all filters in a dense neighborhood (5 x 5 mm or so)
    i. Sweep filter composition (relates to 5i)
    ii. If it matters, try swapopt with different architecture (check if something like ResNet-18 has the right distribution of units in a given layer or set of layers)
3. Swapopt with different local subsampling radii
4. NoLCN swapopt with features from different epochs of model training
5. Implement local subsampling on TPU with ~200 conv1 channels
    1. Consider alternative model architecture with some Retina/LGN like computation up front to filters in V1-analogue to be more OS
